uspace/lib/c/generic/thread/tls.c

```diff
@@ -127,6 +131,26 @@ static tcb_t *tls_make_generic(const void *elf, void *(*alloc)(size_t, size_t))
 
 	uintptr_t bias = elf_get_bias(elf);
 
+	/*
+	 * FIXME: I couldn't convince myself this is correct, but I couldn't
+	 * find a case where it breaks either: if the actual alloc_size is
+	 * bigger than the requested size (tls->p_memsz), the alignment padding
+	 * will be placed at the beginning (because TLS is (at least sometimes?)
+	 * indexed with negative offsets from the TCB pointer).
+	 *
+	 * Now we will copy the initialization data to a position at the start of
+	 * the allocation, so if the padding has nonzero size, if think the initialization
+	 * data is now incorrectly offset by its size?
+	 *
+	 * Maybe a diagram helps explaining this?
+	 * |    allocation   |     |
+	 * | paddding | data | tcb |
+	 *   ^
+	 *   +--- we will copy the initialization data here
+	 *              ^
+	 *              +--- but the data should be actually here?
+	 */
+
```

Comment:
The diagram is incorrect. The padding of the TLS section in variant 2 is between data and tcb. This more or less directly follows from the offset calculations in [https://www.akkadia.org/drepper/tls.pdf](https://www.akkadia.org/drepper/tls.pdf).

Reply:
Also, just common sense. Padding exists to make things aligned. `data` needs to be aligned to TLS alignment, and TCB needs to be aligned to whatever TCB alignment is.

So, in variant II, you allocate memory aligned to `tls_align` and `alignof(tcb_t)`, and add just enough padding that the start of `tcb_t` is aligned correctly within the allocated memory.

Reply:
Now that I'm looking at it, I actually think the calculation in `modules_process_tls()` is incorrect. The old comment over it definitely is.

Reply:
aha, welp, I based a lot of my understanding on the `modules_process_tls` comment :D if that is wrong, then I guess I'm wrong on many things...

Reply:
> So, in variant II, you allocate memory aligned to both tls_align and TCB alignment, and add just enough padding after TLS data that the start of tcb_t is aligned correctly within the allocated memory.

but the accesses to TLS are made as offsets from the `tcb_t` pointer, right? so if we _"add just enough padding after TLS data"_ to align this pointer to whatever we want, then this could break the alignment of the data in TLS, no?